### PR TITLE
Cleaned up container_traits.

### PR DESCRIPTION
**Pull request by octopus-prime**
Cleaned up push_back_container and append_container.

Removed reserve optimization in append_container since we must provide reserve for different container types (e.g. string and deque) and the container implementation should decide how to append elements as fast as possible.

### CODE REVIEW

**Comment by djowel** [comment]
Oops. We need that optimization... please to find another way.

**Comment by octopus-prime** [comment]
Hm... How to provide reserve for so many container types?

string, vector, deque, unordered_map, unordered_multimap, unordered_set, unordered_multiset...

**Comment by octopus-prime** [comment]
Reserve optimization for most important types.

**Comment by djowel** [comment]
yes, we'll have to do some special cases here. that's necessary for optimization.

**Comment by octopus-prime** [comment]
Do you have some simple grammars for testing container traits using append?

**Comment by hkaiser** [comment]
@octopus-prime What's your rationale for this change? Shouldn't we rather provide specializations for those container types you're mentioning?

**Comment by octopus-prime** [comment]
@hkaiser Look at pull request #169. There we removed specializations from push_back_container and used the insert() method witch all common containers provide. So we have less code and we provide more container types - out of the box.

The basic idea in pull request #170 was the same for append_container. To remove specializations and use the insert() method witch all common containers provide. You know providing specializations for each container type on earth is endless work with a lot of spaghetti specialization code...

But djowel told me that we need the reserve() optimization. So the the basic idea is gone...
We need specializations. Perhaps we can close this pull request.

Nevertheless, we should have tests for the append_container. So i need simple grammars for testing container traits using append.

**Comment by djowel** [comment]
The reserve optimization *IS* important. The tricky part is how to generalize this for all containers. For example, we can use a metaprogram that detects the availability of the reserve member function for a container C.

**Comment by octopus-prime** [comment]
Done. Please do an intense code review!!

**Comment by djowel** [comment]
Where's the "like" button? Boost folks are smart! :-)

File: include/boost/spirit/home/x3/support/traits/container_traits.hpp [hunk 1]
```diff
@@ -155,27 +165,21 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
...
+            c.reserve(size);
         }
 
         template <typename Iterator>
         static bool call(Container& c, Iterator first, Iterator last)
         {
-            reserve(c, c.size() + std::distance(first, last));
-            insert(c, first, last);
+        	reserve(c, c.size() + std::distance(first, last), std::integral_constant<bool, detail::has_reserve_method<Container>::value>());
+        	c.insert(c.end(), first, last);
```
Comment (tomilov) [inline include/boost/spirit/home/x3/support/traits/container_traits.hpp]: c.end() returns iterator, c.insert accept const_iterator. There is unneeded conversion.

File: include/boost/spirit/home/x3/support/traits/container_traits.hpp [hunk 2]
```diff
@@ -155,27 +165,21 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
...
-            std::copy(first, last, std::inserter(c, c.end()));
+            c.reserve(size);
         }
 
         template <typename Iterator>
         static bool call(Container& c, Iterator first, Iterator last)
         {
-            reserve(c, c.size() + std::distance(first, last));
-            insert(c, first, last);
+        	reserve(c, c.size() + std::distance(first, last), std::integral_constant<bool, detail::has_reserve_method<Container>::value>());
```
Comment (tomilov) [inline include/boost/spirit/home/x3/support/traits/container_traits.hpp]: unsigned + signed. Is there warning issue?

Comment (tomilov) [inline include/boost/spirit/home/x3/support/traits/container_traits.hpp]: It better to use `typename detail::has_reserve_method<Container>::type{}` then `std::integral_constant<bool, detail::has_reserve_method<Container>::value>()`.

**Comment by octopus-prime** [comment]
@tomilov Thanks!

- c.cend() ... done.
- static_cast< std::size_t >(std::distance(first, last)) ... i am not sure.
- typename detail::has_reserve_method<Container>::type{} ... looks very nice! done.

**Comment by tomilov** [comment]
Defenitely there is an issue with multipass guarantee absence for input iterators. For them you cannot call distance, then read the same range once again. One can get logical (or even hard) error when using std::vector+std::istream_iterator combination.

**Comment by octopus-prime** [comment]
From Qi documentation:

Backtracking in Spirit.Qi requires the use of the following types of iterator: forward, bidirectional, or random access. Because of backtracking, input iterators cannot be used. Therefore, the standard library classes std::istreambuf_iterator and std::istream_iterator, that fall under the category of input iterators, cannot be used.

Lack of X3 documentation here, but I think the same for X3.

**Comment by octopus-prime** [comment]
If you try to use std::istream_iterator<char> as iterator type in x3::parse() you get:

error: conversion from ‘std::input_iterator_tag’ to non-scalar type ‘std::forward_iterator_tag’ requested

**Comment by octopus-prime** [comment]
Some progress in merging this pull request?

**Comment by djowel** [comment]
allow me to do another review this weekend.

**Comment by octopus-prime** [comment]
In container_traits.hpp we have

template <typename Container, typename T>
inline bool push_back(Container& c, T&& val)

and

template <typename T>
inline bool push_back(unused_type, T const&)

I think there should be symmetry. So we should change the latter to

template <typename T>
inline bool push_back(unused_type, T&&)

?

Currently we can't do

x3::traits::push_back(attr, std::make_pair(key, value));

if attr is an unused_type.

**Comment by djowel** [comment]
Unless there is an objection (Hartmut?), I think this is good to go.

**Comment by octopus-prime** [comment]
Using end() because I got compiler errors with gcc 5.3 on Ubuntu 14.04.
The include files there want still iterator and not const_iterator.
end() should discard any of these problems.

**Comment by hkaiser** [comment]
LGTM!

**Comment by octopus-prime** [comment]
gcc 5.2.1 on Ubuntu 15.10 can build.
gcc 5.3.0 on Ubuntu 14.04 can build.

**Pull request merged**